src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      // This is a synchronous resolve that should fail.
+      ModuleRegistry::resolve(js, "file:///foo", "default"_kjc);
+      KJ_FAIL_ASSERT("Should have failed");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: The module evaluation did not complete synchronously. "
+                       "This is not permitted for synchronous require(...). "
+                       "Use await import(...) instead.");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Awaiting a never resolved promise in synchronous require fails as expected") {
+  PREAMBLE([&](Lock& js) {
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+    auto foo = kj::str("const p = new Promise(() => {}); await p;");
+    bundleBuilder.addEsmModule("foo", foo.slice(0, foo.size()).attach(kj::mv(foo)));
+
+    auto bar = kj::str("export default 123;");
+    bundleBuilder.addEsmModule("bar", bar.slice(0, bar.size()).attach(kj::mv(bar)));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      // This is a synchronous resolve that should successfully wait on the async
+      // import for the bar module.
+      ModuleRegistry::resolve(js, "file:///foo", "default"_kjc);
+      KJ_FAIL_ASSERT("Should have failed");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: The module evaluation did not complete synchronously. "
+                       "This is not permitted for synchronous require(...). "
+                       "Use await import(...) instead.");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Throwing an exception inside a ESM module works as expected") {
+  PREAMBLE([&](Lock& js) {
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+    auto foo = kj::str("throw new Error('foo');");
+    bundleBuilder.addEsmModule("foo", foo.slice(0, foo.size()).attach(kj::mv(foo)));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      ModuleRegistry::resolve(js, "file:///foo", "default"_kjc);
+      JSG_FAIL_REQUIRE(Error, "Should have thrown");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: foo");
+    });
+
+    js.tryCatch([&] {
```

Comment:
Nit: If this block is repeated on purpose, add a comment stating this. This will make it clear from the start that the blocks are identical. Would a 2-iteration loop be an option here?

Reply:
Actually,..... I'm not sure why it ended up being repeated lol